File: quantum-sensing-jl/dynamics.jl
# CPU dynamics — Lindblad evolution for MZI, two-arm, and single-mode sensing.

export get_state_mzi_heis, get_state_arms, get_state_single_mode, apply_bs

"""
Mach-Zehnder interferometer: Heisenberg picture formulation.

The density matrix remains fixed in the computational basis (C, D modes).
Beam splitters are encoded as operator transformations:
  a_A = (a_C + a_D)/√2  (arm A)
  a_B = (a_C - a_D)/√2  (arm B)

Evolution sequence:
1. Input noise: Lindblads √κ a_C, √κ a_D act on ρ
2. Arm dynamics: H and Lindblads expressed in arm operators {a_A, a_B}
3. Output noise: Lindblads √κ a_C, √κ a_D act on ρ
"""
function get_state_mzi_heis(;
    Δ::Real=0.0,
    ϵ_a::Real=0.0,
    ϵ_p::Real=0.0,
    η_arms::Tuple{Real, Real}=(1.0, 1.0),
    pn_arms::Tuple{Real, Real}=(0.0, 0.0),
    η_in::Tuple{Real, Real}=(1.0, 1.0),
    pn_in::Tuple{Real, Real}=(0.0, 0.0),
    η_out::Tuple{Real, Real}=(1.0, 1.0),
    pn_out::Tuple{Real, Real}=(0.0, 0.0),
    t_final::Real=1.0,
    ρ=nothing,
    N_basis::Int=20,
    sensing_mode::Symbol=:single,
    show_progress=false
)
    # Computational Basis Operators (C, D modes)
    aC = tensor(destroy(N_basis), qeye(N_basis))
    aD = tensor(qeye(N_basis), destroy(N_basis))
    NC = aC' * aC
    ND = aD' * aD

    # Arm Basis Operators
    aA = (aC + aD) / √2
    aB = (aC - aD) / √2
    NA = aA' * aA
    NB = aB' * aB

    # Input State
    if isnothing(ρ)
        ψ_vac = tensor(fock(N_basis, 0), fock(N_basis, 0))
        ρ = ket2dm(ψ_vac)
    else
        ρ = isket(ρ) ? ket2dm(ρ) : ρ
    end

    # Noise Application Helper
    function apply_noise_stage(ρ_curr, η_vals, pn_vals, ops, N_ops)
        L_stage = QuantumObject[]
        t_stage = 1.0

        for (i, η) in enumerate(η_vals)
            if η < 1.0
                κ = loss_to_kappa(1 - η, t_stage)
                push!(L_stage, √κ * ops[i])
            end
        end

        for (i, pn) in enumerate(pn_vals)
            if pn > 0.0
                χ = phirms_to_chi(pn, t_stage)
                push!(L_stage, √χ * N_ops[i])
            end
        end

        if isempty(L_stage)
            return ρ_curr
        else
            res = mesolve(0*ops[1], ρ_curr, [0, t_stage], L_stage,
                         progress_bar=Val(false))
            return res.states[end]
        end
    end

    # Stage 1: Input Noise (C, D operators)
    ρ = apply_noise_stage(ρ, η_in, pn_in, [aC, aD], [NC, ND])

    # Stage 2: Arm Dynamics
    if sensing_mode == :single
        H = Δ * NA + ϵ_a * (aA' + aA) + 1im * ϵ_p * (aA' - aA)
    elseif sensing_mode == :differential
        H = Δ * (NA - NB) +
            ϵ_a * ((aA' + aA) - (aB' + aB)) +
            1im * ϵ_p * ((aA' - aA) - (aB' - aB))
    else
        error("sensing_mode must be :single or :differential")
    end

    L_arms = QuantumObject[]
    if η_arms[1] < 1.0
        κA = loss_to_kappa(1 - η_arms[1], t_final)
        push!(L_arms, √κA * aA)
    end
    if η_arms[2] < 1.0
        κB = loss_to_kappa(1 - η_arms[2], t_final)
        push!(L_arms, √κB * aB)
    end
    if pn_arms[1] > 0.0
        χA = phirms_to_chi(pn_arms[1], t_final)
        push!(L_arms, √χA * NA)
    end
    if pn_arms[2] > 0.0
        χB = phirms_to_chi(pn_arms[2], t_final)
        push!(L_arms, √χB * NB)
    end

    res_arms = mesolve(H, ρ, range(0, t_final, length=2), L_arms,
                       progress_bar=Val(show_progress))
    ρ = res_arms.states[end]

    # Stage 3: Output Noise (C, D operators)
    ρ = apply_noise_stage(ρ, η_out, pn_out, [aC, aD], [NC, ND])

    return ρ
end


function get_state_arms(;
    Δ=0.0,
    ηx=1.0,
    ηy=1.0,
    pnx=0.0,
    pny=0.0,
    t_final=1.0,
    N_basis=20,
    ρx=nothing, ρy=nothing, ρ=nothing
)
    if isnothing(ρ)
        ρx0 = isnothing(ρx) ? coherent(N_basis, 1.0) : ρx
        ρy0 = isnothing(ρx) ? basis(N_basis, 0) : ρx
        ρ0 = tensor(ρx0, ρy0)
    else
        ρ0 = ρ
    end
    ρ0 = isket(ρ0) ? ket2dm(ρ0) : ρ0

    ax = tensor(destroy(N_basis), qeye(N_basis))
    ay = tensor(qeye(N_basis), destroy(N_basis))

    # Sensing
    H = Δ * ax' * ax
    t_list = [0.0, t_final]
    result = mesolve(H, ρ0, t_list, progress_bar=false)
    ρ_sensed = result.states[end]

    # Loss and dephasing
    c_ops = QuantumObject[]

    if ηx < 1.0 || ηy < 1.0
        κx = loss_to_kappa(1 - ηx)
        κy = loss_to_kappa(1 - ηy)
        push!(c_ops, sqrt(κx) * ax)
        push!(c_ops, sqrt(κy) * ay)
    end

    if pnx > 0.0 || pny > 0.0
        χx = phirms_to_chi(pnx)
        χy = phirms_to_chi(pny)
        push!(c_ops, sqrt(χx) * ax' * ax)
        push!(c_ops, sqrt(χy) * ay' * ay)
    end

    if !isempty(c_ops)
        H1 = 0.0 * ax' * ax
        result = mesolve(H1, ρ_sensed, t_list, c_ops, progress_bar=false)
        return result.states[end]
    end

    ρ_sensed
end

"""
Single mode state evolution with 3-stage noise pipeline:
  1. Input noise (η_in, pn_in)
  2. Sensing channel (H + channel loss/dephasing/amplitude/phase noise)
  3. Output noise (η_out, pn_out)
"""
function get_state_single_mode(;
    Δ::Real=0.0,
    ϵ_a::Real=0.0,
    ϵ_p::Real=0.0,
    η_in::Real=1.0,
    η_ch::Real=1.0,
    η_out::Real=1.0,
    pn_in::Real=0.0,
    pn_ch::Real=0.0,
    pn_out::Real=0.0,
    σ_a::Real=0.0,
    σ_p::Real=0.0,
    t_final::Real=1.0,
    N_basis::Int=20,
    ρ=nothing
)
    if isnothing(ρ)
        ρ0 = coherent(N_basis, 1.0)
    else
        ρ0 = isket(ρ) ? ket2dm(ρ) : ρ
    end

    a = destroy(N_basis)
    n = a' * a

    # Input noise
    L_in = QuantumObject[]
    t_in = 1.0
    if η_in < 1.0
        κ_in = loss_to_kappa(1 - η_in)
        push!(L_in, sqrt(κ_in) * a)
    end
    if pn_in > 0.0
        χ_in = phirms_to_chi(pn_in)
        push!(L_in, sqrt(χ_in) * n)
    end

    if isempty(L_in)
        ρ1 = ρ0
    else
        res_in = mesolve(0*n, ρ0, [0, t_in], L_in, progress_bar=Val(false))
        ρ1 = res_in.states[end]
    end

    # Sensing Channel
    H = Δ * n + ϵ_a * (a' + a) + 1im * ϵ_p * (a' - a)
    t_list = [0.0, t_final]

    L_ops = QuantumObject[]
    if η_ch < 1.0
        κ = loss_to_kappa(1 - η_ch)
        push!(L_ops, sqrt(κ) * a)
    end
    if pn_ch > 0.0
        χ = phirms_to_chi(pn_ch)
        push!(L_ops, sqrt(χ) * n)
    end
    if σ_a > 0.0
        push!(L_ops, sqrt(σ_a) * (a' + a))
    end
    if σ_p > 0.0
        push!(L_ops, 1im * sqrt(σ_p) * (a' - a))
    end

    res_ch = mesolve(H, ρ1, t_list, L_ops, progress_bar=false)
    ρ_sensed = res_ch.states[end]

    # Output noise
    L_out = QuantumObject[]
    t_out = 1.0
    if η_out < 1.0
        κ_out = loss_to_kappa(1 - η_out)
        push!(L_out, sqrt(κ_out) * a)
    end
    if pn_out > 0.0
        χ_out = phirms_to_chi(pn_out)
        push!(L_out, sqrt(χ_out) * n)
    end

    if isempty(L_out)
        ρout = ρ_sensed
    else
        res_out = mesolve(0*n, ρ_sensed, [0, t_out], L_out, progress_bar=Val(false))
        ρout = res_out.states[end]
    end

    return ρout
end

function apply_bs(ρ_in::QuantumObject, N_basis::Int; reverse::Bool=false, θ=π/4)
    aC = tensor(destroy(N_basis), qeye(N_basis))
    aD = tensor(qeye(N_basis), destroy(N_basis))

    θ = reverse ? -θ : θ
    G_BS = aC' * aD - aD' * aC
    U_BS = exp(θ * G_BS)

    if isket(ρ_in)
        return U_BS * ρ_in
    else
        ρ_arms = U_BS * ρ_in * U_BS'
        return ρ_arms
    end
end

# CPU fallback: accept ops=nothing (GPU ops dispatch goes through the CUDA extension)
get_state_mzi_heis(::Nothing; kwargs...) = get_state_mzi_heis(; kwargs...)
get_state_arms(::Nothing; kwargs...) = get_state_arms(; kwargs...)
get_state_single_mode(::Nothing; kwargs...) = get_state_single_mode(; kwargs...)

File: backaction-qfi/dynamics.py
"""
Quantum state dynamics helpers for sensing — Python port of julia_code/DynamicsHelper.jl (CPU).

Provides MZI Heisenberg-picture evolution, two-arm evolution, single-mode evolution,
and beam-splitter transformations, all using QuTiP mesolve with identical physics
to the Julia CPU module.
"""

import numpy as np
import qutip as qt

from conversions import loss_to_kappa, phirms_to_chi


# ---------------------------------------------------------------------------
# Internal helpers
# ---------------------------------------------------------------------------

def _ensure_dm(state, N_basis=None):
    """Convert ket -> density matrix if needed."""
    if state.isket:
        return qt.ket2dm(state)
    return state


def _apply_noise_stage(rho, eta_vals, pn_vals, ops, n_ops):
    """Apply a loss + dephasing noise stage via mesolve (matches Julia helper)."""
    c_ops = []
    t_stage = 1.0

    for i, eta in enumerate(eta_vals):
        if eta < 1.0:
            kappa = loss_to_kappa(1 - eta, t_stage)
            c_ops.append(np.sqrt(kappa) * ops[i])

    for i, pn in enumerate(pn_vals):
        if pn > 0.0:
            chi = phirms_to_chi(pn, t_stage)
            c_ops.append(np.sqrt(chi) * n_ops[i])

    if len(c_ops) == 0:
        return rho

    H_zero = 0 * ops[0].dag() * ops[0]
    res = qt.mesolve(H_zero, rho, [0, t_stage], c_ops)
    return res.states[-1]


# ---------------------------------------------------------------------------
# MZI Heisenberg-picture evolution  (Julia: get_state_mzi_heis)
# ---------------------------------------------------------------------------

def get_state_mzi_heis(
    *,
    Delta=0.0,
    epsilon_a=0.0,
    epsilon_p=0.0,
    eta_arms=(1.0, 1.0),
    pn_arms=(0.0, 0.0),
    eta_in=(1.0, 1.0),
    pn_in=(0.0, 0.0),
    eta_out=(1.0, 1.0),
    pn_out=(0.0, 0.0),
    t_final=1.0,
    rho=None,
    N_basis=20,
    sensing_mode="single",
):
    """
    Mach-Zehnder interferometer in the Heisenberg picture.

    The density matrix stays in the input common and differential (C, D) basis.
    Beam splitters are encoded as operator transformations:
        a_A = (a_C + a_D)/sqrt(2)   (arm A)
        a_B = (a_C - a_D)/sqrt(2)   (arm B)

    Evolution stages:
        1. Input noise  (Lindblads on C, D)
        2. Arm dynamics (H + Lindblads in A, B basis)
        3. Output noise (Lindblads on C, D)
    """
    # --- IO-basis operators ---
    aC = qt.tensor(qt.destroy(N_basis), qt.qeye(N_basis))
    aD = qt.tensor(qt.qeye(N_basis), qt.destroy(N_basis))
    NC = aC.dag() * aC
    ND = aD.dag() * aD

    # --- Arm-basis operators ---
    aA = (aC + aD) / np.sqrt(2)
    aB = (aC - aD) / np.sqrt(2)
    NA = aA.dag() * aA
    NB = aB.dag() * aB

    # --- Input state ---
    if rho is None:
        psi_vac = qt.tensor(qt.fock(N_basis, 0), qt.fock(N_basis, 0))
        rho_dm = qt.ket2dm(psi_vac)
    else:
        rho_dm = _ensure_dm(rho)

    # --- Stage 1: Input noise (C, D) ---
    rho_dm = _apply_noise_stage(rho_dm, eta_in, pn_in, [aC, aD], [NC, ND])

    # --- Stage 2: Arm dynamics ---
    if sensing_mode == "single":
        H = (Delta * NA
             + epsilon_a * (aA.dag() + aA)
             + 1j * epsilon_p * (aA.dag() - aA))
    elif sensing_mode == "differential":
        H = (Delta * (NA - NB)
             + epsilon_a * ((aA.dag() + aA) - (aB.dag() + aB))
             + 1j * epsilon_p * ((aA.dag() - aA) - (aB.dag() - aB)))
    else:
        raise ValueError("sensing_mode must be 'single' or 'differential'")

    L_arms = []
    if eta_arms[0] < 1.0:
        kappaA = loss_to_kappa(1 - eta_arms[0], t_final)
        L_arms.append(np.sqrt(kappaA) * aA)
    if eta_arms[1] < 1.0:
        kappaB = loss_to_kappa(1 - eta_arms[1], t_final)
        L_arms.append(np.sqrt(kappaB) * aB)
    if pn_arms[0] > 0.0:
        chiA = phirms_to_chi(pn_arms[0], t_final)
        L_arms.append(np.sqrt(chiA) * NA)
    if pn_arms[1] > 0.0:
        chiB = phirms_to_chi(pn_arms[1], t_final)
        L_arms.append(np.sqrt(chiB) * NB)

    res_arms = qt.mesolve(H, rho_dm, np.linspace(0, t_final, 2), L_arms)
    rho_dm = res_arms.states[-1]

    # --- Stage 3: Output noise (C, D) ---
    rho_dm = _apply_noise_stage(rho_dm, eta_out, pn_out, [aC, aD], [NC, ND])

    return rho_dm


# ---------------------------------------------------------------------------
# Two-arm (direct arm basis) evolution  (Julia: get_state_arms)
# ---------------------------------------------------------------------------

def get_state_arms(
    *,
    Delta=0.0,
    eta_x=1.0,
    eta_y=1.0,
    pn_x=0.0,
    pn_y=0.0,
    t_final=1.0,
    N_basis=20,
    rho_x=None,
    rho_y=None,
    rho=None,
):
    """
    Two-arm interferometer: sensing on X arm, then loss + dephasing on both.
    """
    # Input state
    if rho is None:
        rho_x0 = rho_x if rho_x is not None else qt.coherent(N_basis, 1.0)
        rho_y0 = rho_y if rho_y is not None else qt.fock(N_basis, 0)
        rho0 = qt.tensor(rho_x0, rho_y0)
    else:
        rho0 = rho

    rho0 = _ensure_dm(rho0)

    # Operators
    ax = qt.tensor(qt.destroy(N_basis), qt.qeye(N_basis))
    ay = qt.tensor(qt.qeye(N_basis), qt.destroy(N_basis))

    # Sensing
    H = Delta * ax.dag() * ax
    t_list = [0.0, t_final]
    result = qt.mesolve(H, rho0, t_list, [])
    rho_sensed = result.states[-1]

    # Loss and dephasing
    c_ops = []
    if eta_x < 1.0:
        kappa_x = loss_to_kappa(1 - eta_x)
        c_ops.append(np.sqrt(kappa_x) * ax)
    if eta_y < 1.0:
        kappa_y = loss_to_kappa(1 - eta_y)
        c_ops.append(np.sqrt(kappa_y) * ay)
    if pn_x > 0.0:
        chi_x = phirms_to_chi(pn_x)
        c_ops.append(np.sqrt(chi_x) * ax.dag() * ax)
    if pn_y > 0.0:
        chi_y = phirms_to_chi(pn_y)
        c_ops.append(np.sqrt(chi_y) * ay.dag() * ay)

    if len(c_ops) > 0:
        H1 = 0 * ax.dag() * ax
        result1 = qt.mesolve(H1, rho_sensed, t_list, c_ops)
        return result1.states[-1]

    return rho_sensed


# ---------------------------------------------------------------------------
# Single-mode evolution  (Julia: get_state_single_mode)
# ---------------------------------------------------------------------------

def get_state_single_mode(
    *,
    Delta=0.0,
    epsilon_a=0.0,
    epsilon_p=0.0,
    eta_in=1.0,
    eta_ch=1.0,
    eta_out=1.0,
    pn_in=0.0,
    pn_ch=0.0,
    pn_out=0.0,
    sigma_a=0.0,
    sigma_p=0.0,
    t_final=1.0,
    N_basis=20,
    rho=None,
):
    """
    Single-mode channel with 3-stage noise pipeline:
        1. Input noise   (eta_in, pn_in)
        2. Sensing channel  (H + channel loss/dephasing/amplitude/phase noise)
        3. Output noise  (eta_out, pn_out)

    Mirrors Julia DynamicsHelper.get_state_single_mode exactly.
    """
    # Default state
    if rho is None:
        rho0 = qt.ket2dm(qt.coherent(N_basis, 1.0))
    else:
        rho0 = _ensure_dm(rho)

    a = qt.destroy(N_basis)
    n_op = a.dag() * a

    # --- Stage 1: Input noise ---
    L_in = []
    t_in = 1.0
    if eta_in < 1.0:
        kappa_in = loss_to_kappa(1 - eta_in)
        L_in.append(np.sqrt(kappa_in) * a)
    if pn_in > 0.0:
        chi_in = phirms_to_chi(pn_in)
        L_in.append(np.sqrt(chi_in) * n_op)

    if len(L_in) == 0:
        rho1 = rho0
    else:
        H_zero = 0 * n_op
        res_in = qt.mesolve(H_zero, rho0, [0, t_in], L_in)
        rho1 = res_in.states[-1]

    # --- Stage 2: Sensing channel ---
    H = Delta * n_op + epsilon_a * (a.dag() + a) + 1j * epsilon_p * (a.dag() - a)

    L_ch = []
    if eta_ch < 1.0:
        kappa = loss_to_kappa(1 - eta_ch)
        L_ch.append(np.sqrt(kappa) * a)
    if pn_ch > 0.0:
        chi = phirms_to_chi(pn_ch)
        L_ch.append(np.sqrt(chi) * n_op)
    if sigma_a > 0.0:
        L_ch.append(np.sqrt(sigma_a) * (a.dag() + a))
    if sigma_p > 0.0:
        L_ch.append(1j * np.sqrt(sigma_p) * (a.dag() - a))

    res_ch = qt.mesolve(H, rho1, [0.0, t_final], L_ch)
    rho_sensed = res_ch.states[-1]

    # --- Stage 3: Output noise ---
    L_out = []
    t_out = 1.0
    if eta_out < 1.0:
        kappa_out = loss_to_kappa(1 - eta_out)
        L_out.append(np.sqrt(kappa_out) * a)
    if pn_out > 0.0:
        chi_out = phirms_to_chi(pn_out)
        L_out.append(np.sqrt(chi_out) * n_op)

    if len(L_out) == 0:
        return rho_sensed

    H_zero = 0 * n_op
    res_out = qt.mesolve(H_zero, rho_sensed, [0, t_out], L_out)
    return res_out.states[-1]


# ---------------------------------------------------------------------------
# Beam splitter  (Julia: apply_bs)
# ---------------------------------------------------------------------------

def apply_bs(rho_in, N_basis, reverse=False, theta=np.pi / 4):
    """
    Apply a beam-splitter unitary U_BS = exp(theta (a_C^dag a_D - a_D^dag a_C)).

    Works on both kets and density matrices.
    """
    aC = qt.tensor(qt.destroy(N_basis), qt.qeye(N_basis))
    aD = qt.tensor(qt.qeye(N_basis), qt.destroy(N_basis))

    sign = -1 if reverse else 1
    G_BS = aC.dag() * aD - aD.dag() * aC
    U_BS = (sign * theta * G_BS).expm()

    if rho_in.isket:
        return U_BS * rho_in
    else:
        return U_BS * rho_in * U_BS.dag()
